Support Llama 3.2 conversion (text models)

pcuenca opened the pull request:
# What does this PR do?

Update the existing Llama conversion script to support the new 1B and 3B models.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

pcuenca committed:
Support Llama 3.2 conversion (text models)

Co-authored-by: Omar Sanseviero <osanseviero@gmail.com>

pcuenca committed:
Fix rope factor

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_33778). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

pcuenca committed:
Update chat template

Initialize from a well-known template.
The guidance is that the changes should be applied to 3.1 models as
well.

pcuenca commented on src/transformers/models/llama/convert_llama_weights_to_hf.py:
```diff
@@ -149,262 +193,259 @@ def write_model(
+        if instruct:
+            # Guidance from Meta: apply 3.2 changes to 3.1 as well
+            # TODO: 3 should be different (no tool calling)
+            from transformers import AutoTokenizer
+            t = AutoTokenizer.from_pretrained("meta-llama/Llama-3.2-1B-Instruct")
+            additional_kwargs["chat_template"] = t.chat_template
```
Is it acceptable to load from a well-known tokenizer @ArthurZucker? This way they'd always be in sync, and having the very long template in the script is a bit noisy.

ArthurZucker replied:
If we use a revision, fine by me!

pcuenca committed:
Remove import

pcuenca committed:
Support Llama Guard 3 conversion

pcuenca committed:
Tokenizer details

pcuenca committed:
Fix eos added token in base models

pcuenca committed:
Fix generation config for base models

pcuenca committed:
Specify revision for known tokenizers

pcuenca committed:
Style

pcuenca commented on src/transformers/models/llama/convert_llama_weights_to_hf.py:
```diff
@@ -90,7 +93,79 @@
 }
 
-CONTEXT_LENGTH_FOR_VERSION = {"3.1": 131072, "3": 8192, "2": 4096, "1": 2048}
+CONTEXT_LENGTH_FOR_VERSION = {"Guard-3": 131072, "3.2": 131072, "3.1": 131072, "3": 8192, "2": 4096, "1": 2048}
```
The Llama Guard 3 model uses a different rope scale and requires a different chat template. An alternative would be to use `--guard` (which would imply `--instruct`) instead of adding a new version, but I'm not sure if code would be clearer.

pcuenca commented:
This is now ready for review @ArthurZucker. I exported the Llama 3.2 instruct, base and guard models and verified that the configuration files were identical to what's currently in the Hub.

ArthurZucker reviewed (commented):
Thanks, LGTM, what's best is usually to try converting at least llama1 7B, and llama2 7B as well! 🤗 this way we are certain there is not regression

pcuenca committed:
Reuse chat templates for older models

pcuenca committed:
Improve error when converting tokenizer < Llama 3

pcuenca commented:
I converted Llama 2 and only saw a couple of new properties from the default configuration (for example, `mlp_bias`) that look fine to me. The only interesting difference is that the converted weights are now exported in `bfloat16`, while [this repo](https://huggingface.co/meta-llama/Llama-2-7b-chat-hf/tree/main) was exported in `float16`. I believe `bfloat16` is correct, because the [original weights](https://huggingface.co/meta-llama/Llama-2-7b-chat/tree/main) were released with that dtype (and it's consistent with newer models). Let me know if you think it's necessary to override the dtype in this case, but I think it could potentially mess things up when converting fine-tunes exported in the original format.

I found the error message for tokenizer (< 3.x) conversion to be unclear, so I added a explicit reference to requiring `sentencepiece`.

I didn't keep a local copy of the Llama 1 research weights, so I can't test that code path now. I re-requested access to download the weights from Meta, not sure how long it'll take.

ArthurZucker commented:
No worries, LGTM and I think we can merge! 🤗
